### PR TITLE
Windows: restore canvas inside screen bounds

### PR DESCRIPTION
**Pull request by ben-wes**
Changes canvas positioning for Windows users once more according to discussion in https://github.com/pure-data/pure-data/issues/2194.

This replaces the current logic where positions get wrapped with a more predictable solution that keeps the canvas inside screen bounds (these bounds also apply for multi-screen setups and can have a negative origin)

This PR is to be merged into develop branch.

- Closes: https://github.com/pure-data/pure-data/issues/2194

### CODE REVIEW

**Comment by Lucarda** [comment]
@Spacechild1 and @umlaeute i had tested this on Win11 with "single" and "dual" display settings and it all went good.

**Comment by Spacechild1** [comment]
@Lucarda I'll trust you on that :)

**Comment by ben-wes** [comment]
> @Lucarda I'll trust you on that :)

just don't trust me! :) 

(and on a more serious note: thanks! i'm also quite confident that this will improve things after testing with different setups myself)

**Review by umlaeute — CHANGES_REQUESTED** [review]
in https://github.com/pure-data/pure-data/issues/2194#issuecomment-1944226690 you say that
>  removed `$::menubarsize` here since this function is only for windowingsystem "win32" and the value in that case is set to 0 in 


however, this assumption is not true.
the full condition for applying the code  is  https://github.com/pure-data/pure-data/blob/d153c362707b72872c9e9a62a9670a101329c29d/tcl/pdtk_canvas.tcl#L46-L50 which obviously includes macOS (at least for some Tcl/Tk versions), and i'm sure that on macOS the `$::menubarsize` is non-zero.

i think this needs testing on macOS before merging it.

**Comment by ben-wes** [comment]
@umlaeute ah ... damn. the condition is quite obviously not just checking win32. not sure why i missed that. sorry and thanks for checking! i tested on sonoma now and it looks all good. i'd prefer to have a less intrusive change then though and consider `$::menubarsize` just in case.

will test more with an adapted version and update this PR (is that how it works or will we close this one and i'd create another one?)

**Comment by Spacechild1** [comment]
> or will we close this one and i'd create another one?)

No! Just apply the changes in another commit. Or ammend the last commit and force push.

**Comment by ben-wes** [comment]
> Or amend the last commit and force push.

Ok, thanks. Did that now and it should consider the `$::menubarsize` - the problem for me is that it looks good on Win11 and macOS sonoma. But I can't really tell for other setups where this change might be relevant.

**Comment by ben-wes** [comment]
i tested a bit more with a forced `menubarsize` on win11 and sonoma now (where i applied this new logic as well without the conditions just for testing):
* it behaves as expected on win11, respecting an imaginary menubar
  * the menubar is always expected at the top of the whole screen area though - so for screen setups that have a negative origin, the menubar would be expected at `y: -1080` for example. not sure if this is ideal - but the current solution also only takes this variant into account.
  * additionally, the current solution only compensates for a menubar if the restored canvas height is bigger than the available screen height. this means that i should still be able to cover the menubar when `h == height`
* for sonoma, this test doesn't really make sense since it seems like `winfo vrooty` doesn't give negative values there ... at least, i can't properly restore windows at a negative position. this is not a regression though and also my tested scenario with this logic on sonoma doesn't apply in the "real world"

in short: i think, this change makes sense to avoid the disappearing patches for windows users. menubars are considered if present now and the behaviour is more transparent than it was before.

no problem at all though if this PR won't make it and if someone with actual tcl experience can solve this in a better way!

**Comment by ben-wes** [comment]
@umlaeute : this is obviously still not critical at all - but the requested changes should be included (considering the menubar). i rebased to `develop` now and imho, this will potentially avoid confusion (caused by invisible patches) for some Windows users.

**Comment by umlaeute** [comment]
doh, i think i missed the point of the ping: while I approved the changes, i forgot to merge into `develop`

**Comment by ben-wes** [comment]
no problem. this confirmed my point of low priority. :)
anyway: thanks!
